src/main/java/de/zalando/aruha/nakadi/validation/EventBodyMustRespectSchema.java

```diff
@@ -169,7 +165,22 @@ public JSONSchemaValidator(final JSONObject effectiveSchema) {
 
             return Optional.empty();
         } catch (final ValidationException e) {
-            return Optional.of(new ValidationError(e.getMessage()));
+            StringBuilder builder = new StringBuilder();
+            collectErrorMessages(e, builder);
+
+            return Optional.of(new ValidationError(builder.toString()));
+        }
+    }
+
+    private void collectErrorMessages(ValidationException e, StringBuilder builder) {
+        builder.append(e.getMessage());
+
+        if (!e.getCausingExceptions().isEmpty()) {
+            e.getCausingExceptions().stream()
+                    .forEach(causingException -> {
+                        builder.append("\n");
+                        collectErrorMessages(causingException, builder);
```

Comment:
I can only imagine that using an additional data structure, like a stack. Do you think it's possible to do that even without such a data structure?